core/src/main/java/tc/oc/pgm/modules/SpectateMatchModule.java

```diff
@@ -0,0 +1,89 @@
+package tc.oc.pgm.modules;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Multimap;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import org.bukkit.entity.Entity;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
+import org.bukkit.event.Listener;
+import org.bukkit.event.player.PlayerQuitEvent;
+import org.bukkit.event.player.PlayerTeleportEvent;
+import tc.oc.pgm.api.match.Match;
+import tc.oc.pgm.api.match.MatchModule;
+import tc.oc.pgm.api.match.MatchScope;
+import tc.oc.pgm.api.player.MatchPlayer;
+import tc.oc.pgm.community.events.PlayerVanishEvent;
+import tc.oc.pgm.events.ListenerScope;
+import tc.oc.pgm.events.PlayerLeaveMatchEvent;
+
+@ListenerScope(MatchScope.LOADED)
+public class SpectateMatchModule implements MatchModule, Listener {
+
+  private final Match match;
+
+  // Stores which players (the list) is spectating a player(the uuid)
+  private final Multimap<UUID, UUID> spectators = ArrayListMultimap.create();
+
+  public SpectateMatchModule(Match match) {
+    this.match = match;
+  }
+
+  @EventHandler
+  public void onPlayerLeave(PlayerLeaveMatchEvent event) {
+    event.getPlayer().getSpectators().forEach(p -> p.getBukkit().setSpectatorTarget(null));
+  }
+
+  @EventHandler
+  public void onPlayerQuit(PlayerQuitEvent event) {
+    spectators.values().removeIf(event.getPlayer().getUniqueId()::equals);
+  }
+
+  @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
+  public void onPlayerVanish(PlayerVanishEvent event) {
+    if (!event.isVanished()) return;
+    event.getPlayer().getSpectators().forEach(p -> p.getBukkit().setSpectatorTarget(null));
+  }
+
+  @EventHandler
+  public void onPlayerTeleport(PlayerTeleportEvent event) {
+    if (event.getCause() == PlayerTeleportEvent.TeleportCause.UNKNOWN) {
+      MatchPlayer player = match.getPlayer(event.getPlayer());
+      if (player != null) {
+        MatchPlayer spectating = player.getSpectatorTarget();
+        if (spectating != null) { // Player is going into spectate
+          spectators.get(spectating.getId()).add(player.getId());
```

Comment:
I think you should normally use https://guava.dev/releases/23.0/api/docs/com/google/common/collect/Multimap.html#put-K-V-